tools/canonicalize-proto-file/Main.hs

```diff
@@ -11,28 +10,110 @@
 {-# LANGUAGE TypeFamilies           #-}
 {-# LANGUAGE TypeOperators          #-}
 
-module Main (main) where
-
-import           Control.Monad.Except
-import           Data.List                        (sort, sortOn)
-import qualified Data.List.NonEmpty              as NE
-import           Data.RangeSet.List               (fromRangeList, toRangeList)
-import           Data.Semigroup                   (Min(..))
-import           Options.Generic
-import           Prelude                          hiding (FilePath)
-import           Proto3.Suite.DotProto.AST
-import           Proto3.Suite.DotProto.Generate
-import           Proto3.Suite.DotProto.Rendering
-import           Proto3.Wire.Types                (FieldNumber (..))
-import           Turtle                           (FilePath)
+module Main where
+
+import Control.Monad (guard)
+import Control.Monad.Except (runExceptT)
+
+import Data.List (sort, sortOn)
+import Data.List.NonEmpty qualified as NE
+import Data.Semigroup (Min(..))
+
+import GHC.Generics (Generic)
+
+import Options.Generic 
+  ( ParseRecord
+  , Unwrapped
+  , Wrapped 
+  , (:::)
+  , type (<?>) (..)
+  , unwrapRecord
+  )
+
+import Prelude hiding (FilePath)
+
+import Proto3.Suite.DotProto.AST 
+  ( DotProto (..)
+  , DotProtoDefinition (..)
+  , DotProtoEnumPart (..)
+  , DotProtoEnumValue 
+  , DotProtoField (..)
+  , DotProtoIdentifier (..)
+  , DotProtoImport (..)
+  , DotProtoMessagePart (..)
+  , DotProtoOption (..)
+  , DotProtoPackageSpec (..)
+  , DotProtoServicePart (..)
+  , DotProtoReservedField (..)
+  , DotProtoType (..)
+  , DotProtoValue (..)
+  , Path (..)
+  , RPCMethod (..)
+  )
+import Proto3.Suite.DotProto.Generate (readDotProtoWithContext)
+import Proto3.Suite.DotProto.Rendering (defRenderingOptions, toProtoFile)
+import Proto3.Wire.Types (FieldNumber (..))
+
+import Turtle (FilePath)
+
+--------------------------------------------------------------------------------
 
 data Args w = Args
   { includeDir :: w ::: [FilePath] <?> "Path to search for included .proto files (can be repeated, and paths will be searched in order; the current directory is used if this option is not provided)"
   , proto      :: w ::: FilePath   <?> "Path to input .proto file"
   } deriving Generic
+
 instance ParseRecord (Args Wrapped)
+
 deriving instance Show (Args Unwrapped)
 
+-- | Check if the given values is some given bounds (inclusive).
+between :: (Ord a, Enum a) => a -> (a, a) -> Bool
+between x i = fst i <= x && x <= snd i
+
+-- | @('isOverlappingIntervals' a b :: 'Bool')@ is a relation between two 
+-- interval-like types @a@ and @b@ that is 'True' when the two intervals are 
```

Comment:
I think the word `types` is confusing here because `a` and `b` are terms, not types.